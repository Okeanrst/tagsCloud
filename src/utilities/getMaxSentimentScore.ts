import { TagDataT } from 'types/types';

export const getMaxSentimentScore = (tagsData: ReadonlyArray<TagDataT>) => {
  return Math.max(-Infinity, ...tagsData.map(({ sentimentScore }) => sentimentScore));
};
