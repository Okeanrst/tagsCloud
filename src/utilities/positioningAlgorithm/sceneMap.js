import IntersectionError from './IntersectionError';

const X = 'x';
const MINUS_X = '-x';
const Y = 'y';
const MINUS_Y = '-y';

function sceneMap() {
	this.sceneMap = {
		'xy': [],
		'x-y': [],
		'-x-y': [],
		'-xy': []
	};
	this.sizeX = 0;
	this.sizeMinusX = 0;
	this.sizeY = 0;
	this.sizeMinusY = 0;	
}

function calcQuarter(xIsPositive, yIsPositive) {
	return `${xIsPositive ? '' : '-'}x${yIsPositive ? '' : '-'}y`;
}

sceneMap.prototype.getSceneSize = function getSceneSize() {
	const {sizeX, sizeMinusX, sizeY, sizeMinusY} = this;	
	return {[X]: sizeX, [MINUS_X]: sizeMinusX, [Y]: sizeY, [MINUS_Y]: sizeMinusY};
}

sceneMap.prototype.calcSceneSize = function getSceneSize() {
  //Y
	if (this.sceneMap['xy'].length > this.sizeY) {
    this.sizeY = this.sceneMap['xy'].length;
	}

  if (this.sceneMap['-xy'].length > this.sizeY) {
    this.sizeY = this.sceneMap['-xy'].length;
  }
	//-Y
  if (this.sceneMap['x-y'].length > this.sizeMinusY) {
    this.sizeMinusY = this.sceneMap['x-y'].length
  }
  if (this.sceneMap['x-y'].length > this.sizeMinusY) {
    this.sizeMinusY = this.sceneMap['x-y'].length
  }

  //X
  this.sceneMap['xy'].forEach(row => {
    if (row.length > this.sizeX) {
      this.sizeX = row.length;
    }
  });
  this.sceneMap['x-y'].forEach(row => {
    if (row.length > this.sizeX) {
      this.sizeX = row.length;
    }
  });
	//-X
  this.sceneMap['-xy'].forEach(row => {
    if (row.length > this.sizeMinusX) {
      this.sizeMinusX = row.length;
    }
  });
  this.sceneMap['-x-y'].forEach(row => {
    if (row.length > this.sizeMinusX) {
      this.sizeMinusX = row.length;
    }
  });
}

sceneMap.prototype._setDataAtPosition = function setDataAtPosition(x, y, val = true) {
  const xIsPositive = x >= 0;
  const yIsPositive = y >= 0;
  const quarter = calcQuarter(xIsPositive, yIsPositive);
  const row = Math.abs(y);
  const col = Math.abs(x);
  if (!Array.isArray(this.sceneMap[quarter][Math.abs(y)])) {
    this.sceneMap[quarter][row] = [];
  }
  this.sceneMap[quarter][row][col] = val;
}

sceneMap.prototype.setDataAtPosition = function setDataAtPosition(x, y, val = true) {
	if (typeof x !== 'number' || typeof y !== 'number') {
		throw new Error('setDataAtPosition error: typeof x !== number || typeof y !== number');
	}
	if (x === 0 || y === 0) {
		throw new Error('setDataAtPosition error: x === 0 || y === 0');
	}
	if (this.getDataAtPosition(x, y)) {
		throw new IntersectionError();
	}
  this._setDataAtPosition(x, y);
}

sceneMap.prototype.releasePosition = function(x, y) {
  this._setDataAtPosition(x, y, undefined);
}

sceneMap.prototype.getDataAtPosition = function getDataAtPosition(x, y) {
	if (typeof x !== 'number' || typeof y !== 'number') {
		throw new Error('getDataAtPosition error: typeof x !== number || typeof y !== number');
	}
	if (x === 0 || y === 0) {
		throw new Error('getDataAtPosition error: x === 0 || y === 0');
	}	
	const quarter = calcQuarter(x >= 0, y >= 0);
	const row = Math.abs(y);
	const col = Math.abs(x);
	if (!Array.isArray(this.sceneMap[quarter][row])) {
		return;
	}	
	return this.sceneMap[quarter][row][col];
}

sceneMap.prototype.drawItself = function drawItself() {
	//сверху вниз
	const sceneSize = this.getSceneSize();
	const topRow = sceneSize[Y];
	const bottomRow = -sceneSize[MINUS_Y];
	const leftCol = -sceneSize[MINUS_X];
	const rightCol = sceneSize[X];
	
	let res = '';	

	for (let row = topRow; row >= bottomRow; row--) {				
		for (let col = leftCol; col <= rightCol; col++) {
			if (row === 0) {
				res +=  '-';				
			} else {
				res += col === 0 ? '|' : this.getDataAtPosition(col, row) ? '#' : '.';
			}
		}
		res += '\n';
	}

	console.log(res, '\n\n');	
}

sceneMap.rectSizeToPositionUnits = function (rectSize, ratio) {
	return Math.ceil(rectSize / ratio);
}

sceneMap.rectToSceneMap = function rectToSceneMap(rectVal, ratio) {
	return Math.sign(rectVal)*Math.ceil(Math.abs(rectVal / ratio));
}

sceneMap.sceneMapToRect = function sceneMapToRect(mapVal, ratio) {
	return mapVal * ratio;
}

sceneMap.countPositions = function(begin, end) {
	if (begin > end) {
    throw new Error('countPositions error: can not be begin > end');
	}
  if (begin === 0 || end === 0) {
    throw new Error('countPositions error: begin, end can not be zero');
  }
	return begin < 0 && end > 0 ? end - begin : end - begin + 1;
}

sceneMap.takePositionsFromFirst = function(first, count) {
  if (count <= 0) throw new Error('"count" must be positive');
  if (first === 0) throw new Error('"first" can not be zero');
  if (count === 1) return first;
  let res = first + count;
  if (!(first < 0 && res >= 0)) {
    res--;
  }
  return res;
}

sceneMap.takePositionsFromLast = function(last, count) {
  if (count <= 0) throw new Error('"count" must be positive');
  if (last === 0) throw new Error('"last" can not be zero');
  if (count === 1) return last;

  let res = last - count;
  if (res >= 0 || last < 0) {
    res++;
  }
  return res;
}

sceneMap.nextPosition = cur => cur === -1 ? cur + 2 : cur + 1;
sceneMap.prevPosition = cur => cur === 1 ? cur - 2 : cur - 1;
sceneMap.changePosition = (cur, diff) => {
  if (cur === 0) throw new Error('"cur" can not be zero');
  let res = cur + diff;
  if (cur > 0 && res <= 0) {
    res--;
  }
  if (cur < 0 && res >= 0) {
    res++;
  }
  return res;
}

export default sceneMap;
export const dimensions = {X, MINUS_X, MINUS_Y, Y};