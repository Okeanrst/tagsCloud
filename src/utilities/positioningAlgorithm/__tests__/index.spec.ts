import { OPEN_SANS_FONT } from 'constants/index';
import { prepareTagsData } from 'utilities/tagsCloud/tagsCloud';
import { formRectAreaMapKey, prepareRectAreasMaps } from 'utilities/prepareRectAreasMaps';
import { getMaxSentimentScore } from 'utilities/tagsCloud/getMaxSentimentScore';
import {
  calcTagsPositions,
  isVacancyLargeEnoughToFitRect,
  rotateRectArea,
  releaseRectAreaPositionsOnSceneMap,
} from '../calcTagsPositions';
import { PositionT, SceneMap } from '../sceneMap';
import entryData from './entryData.json';
import preparedData from './preparedData.json';
import tagsRectAreasMaps from './rectAreasMaps.json';
import fullSizeFilledRectAreasMaps from './fullSizeFilledRectAreasMaps.json';

jest.mock('utilities/common/getRandomRGBColor', () => {
  return {
    getRandomRGBColor: () => {
      return 'rgb(155, 155, 155)';
    },
  };
});

jest.mock('utilities/getGlyphsMap', () => {
  return {
    ...jest.requireActual('utilities/getGlyphsMap'),
    getRectAreaMap: () => {
      return {};
    },
  };
});

describe('positioningAlgorithm tests', () => {
  describe('prepareData tests', () => {
    it(`should return correct result`, () => {
      const maxSentimentScore = getMaxSentimentScore(entryData);
      expect(prepareTagsData(entryData, {minFontSize: 6, maxFontSize: 36, maxSentimentScore})).toMatchSnapshot();
    });
  });

  describe('prepareRectAreasMaps tests', () => {
    it(`should return array with correct value (right shape)`, async () => {
      const tagsRectAreasMaps = await prepareRectAreasMaps(preparedData, { resolution: 2, fontFamily: OPEN_SANS_FONT });
      expect(tagsRectAreasMaps).toStrictEqual(
        preparedData.map(({ label, fontSize }) => ({ key: formRectAreaMapKey(label, fontSize), map: null, mapMeta: null })),
      );
    });
  });

  describe('calcTagsPositions tests', () => {
    it(`return result is equal to snapshot`, async () => {
      const { tagsPositions } = await calcTagsPositions(preparedData, tagsRectAreasMaps, [], {sceneMapResolution: 2});
      expect(tagsPositions).toMatchSnapshot();
    });

    it(`no intersection`, async () => {
      const { tagsPositions } = await calcTagsPositions(preparedData, fullSizeFilledRectAreasMaps, [], {sceneMapResolution: 2});

      let intersection = 0;
      const map = new Map();
      tagsPositions.forEach(({ rectTop, rectBottom, rectRight, rectLeft }) => {
        const firstRow = SceneMap.getPrevPositionFromEdge(rectTop);
        const lastRow = SceneMap.getNextPositionFromEdge(rectBottom);
        const firstCol = SceneMap.getNextPositionFromEdge(rectLeft);
        const lastCol = SceneMap.getPrevPositionFromEdge(rectRight);
        for (let row = firstRow; row >= lastRow; row--) {
          for (let col = firstCol; col <= lastCol; col++) {
            const key = `${row},${col}`;
            if (map.has(key)) {
              intersection++;
            } else {
              map.set(key, true);
            }
          }
        }
      });
      expect(intersection).toBe(0);
    });
  });

  describe('isVacancyLargeEnoughToFitRect tests', () => {
    it(`should return correct result`, () => {
      const isLargeEnough = isVacancyLargeEnoughToFitRect({rows: 2, cols: 4}, {top: 4, bottom: 3, left: -Infinity, right: 5});
      expect(isLargeEnough).toBe(true);
    });
    it(`should return correct result (1)`, () => {
      const isLargeEnough = isVacancyLargeEnoughToFitRect({rows: 1, cols: 1}, {top: -1, bottom: -Infinity, left: -5, right: -5});
      expect(isLargeEnough).toBe(true);
    });
    it(`should return correct result (2)`, () => {
      const isLargeEnough = isVacancyLargeEnoughToFitRect({rows: 5, cols: 2}, {top: -1, bottom: -Infinity, left: -5, right: Infinity});
      expect(isLargeEnough).toBe(true);
    });
    it(`should return correct result (3)`, () => {
      const isLargeEnough = isVacancyLargeEnoughToFitRect({rows: 100, cols: 2000}, {right: Infinity, left: -Infinity, bottom: 2, top: Infinity});
      expect(isLargeEnough).toBe(true);
    });
    it(`should return falsy result for rotated vacancy (1)`, () => {
      const isLargeEnough = isVacancyLargeEnoughToFitRect({rows: 3, cols: 4}, {right: 4, left: -Infinity, bottom: 2, top: 3});
      expect(isLargeEnough).toBe(false);
    });
  });

  describe('rotateRectArea tests', () => {
    it(`should return correct result`, () => {
      expect(rotateRectArea({rows: 5, cols: 2})).toStrictEqual({rows: 2, cols: 5});
    });
  });

  describe('releaseRectAreaPositionsOnSceneMap tests', () => {
    it(`should return correct result`, () => {
      const sceneMapPositions: PositionT[] = [
        [1, 1], [2, 1], [3, 1], [4, 1], [5, 1], [6, 1],
        [1, 2], [2, 2], [3, 2], [4, 2], [5, 2], [6, 2],
        [1, 3], [2, 3], [3, 3], [4, 3], [5, 3], [6, 3],
        [1, 4], [2, 4], [3, 4], [4, 4], [5, 4], [6, 4],
      ];
      const tagPosition = {bottom: 2, top: 4, left: 3, right: 6, rotate: false };
      const rectAreaMap = [[true, false, true, false], [true, true, true, true], [true, true, true, true]];
      const sceneMap = releaseRectAreaPositionsOnSceneMap(sceneMapPositions, tagPosition, rectAreaMap);
      expect(sceneMap.toPositions()).toStrictEqual([[1,4],[2,4],[4,4],[6,4],[1,3],[2,3],[1,2],[2,2],[1,1],[2,1],[3,1],[4,1],[5,1],[6,1]]);
    });
  });
});
