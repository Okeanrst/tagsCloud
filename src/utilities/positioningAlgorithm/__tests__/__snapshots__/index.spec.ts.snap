// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`positioningAlgorithm tests calcTagsPositions tests return result is equal to snapshot 1`] = `
Array [
  Object {
    "bottom": -7,
    "color": "rgb(155, 155, 155)",
    "fontSize": 36,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__Mixes",
    "label": "Mixes",
    "left": -24,
    "rectBottom": -20,
    "rectLeft": -50,
    "rectRight": 48,
    "rectTop": 16,
    "right": 24,
    "rotate": false,
    "sentimentScore": 895,
    "square": 672,
    "top": 7,
  },
  Object {
    "bottom": -25,
    "color": "rgb(155, 155, 155)",
    "fontSize": 11,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__legendary nightclub",
    "label": "legendary nightclub",
    "left": -29,
    "rectBottom": -50,
    "rectLeft": -58,
    "rectRight": -46,
    "rectTop": 52,
    "right": -25,
    "rotate": true,
    "sentimentScore": 150,
    "square": 255,
    "top": 26,
  },
  Object {
    "bottom": 8,
    "color": "rgb(155, 155, 155)",
    "fontSize": 9,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__Panorama Bar in Berlin",
    "label": "Panorama Bar in Berlin",
    "left": -24,
    "rectBottom": 12,
    "rectLeft": -48,
    "rectRight": 50,
    "rectTop": 22,
    "right": 25,
    "rotate": false,
    "sentimentScore": 83,
    "square": 196,
    "top": 11,
  },
  Object {
    "bottom": -58,
    "color": "rgb(155, 155, 155)",
    "fontSize": 8,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__Kantine am Berghain",
    "label": "Kantine am Berghain",
    "left": -25,
    "rectBottom": -116,
    "rectLeft": -50,
    "rectRight": -42,
    "rectTop": -36,
    "right": -22,
    "rotate": true,
    "sentimentScore": 59,
    "square": 160,
    "top": -19,
  },
  Object {
    "bottom": 12,
    "color": "rgb(155, 155, 155)",
    "fontSize": 9,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__Marcel Dettmann",
    "label": "Marcel Dettmann",
    "left": -24,
    "rectBottom": 20,
    "rectLeft": -48,
    "rectRight": 26,
    "rectTop": 30,
    "right": 13,
    "rotate": false,
    "sentimentScore": 87,
    "square": 148,
    "top": 15,
  },
  Object {
    "bottom": -54,
    "color": "rgb(155, 155, 155)",
    "fontSize": 8,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__San Soda's Panorama Bar",
    "label": "San Soda's Panorama Bar",
    "left": -21,
    "rectBottom": -108,
    "rectLeft": -44,
    "rectRight": -36,
    "rectTop": -12,
    "right": -19,
    "rotate": true,
    "sentimentScore": 50,
    "square": 144,
    "top": -7,
  },
  Object {
    "bottom": 16,
    "color": "rgb(155, 155, 155)",
    "fontSize": 8,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__Berghain resident",
    "label": "Berghain resident",
    "left": -24,
    "rectBottom": 30,
    "rectLeft": -48,
    "rectRight": 20,
    "rectTop": 38,
    "right": 10,
    "rotate": false,
    "sentimentScore": 73,
    "square": 136,
    "top": 19,
  },
  Object {
    "bottom": -63,
    "color": "rgb(155, 155, 155)",
    "fontSize": 8,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__Quantified Drunk",
    "label": "Quantified Drunk",
    "left": -19,
    "rectBottom": -126,
    "rectLeft": -38,
    "rectRight": -30,
    "rectTop": -60,
    "right": -16,
    "rotate": true,
    "sentimentScore": 50,
    "square": 132,
    "top": -31,
  },
  Object {
    "bottom": 20,
    "color": "rgb(155, 155, 155)",
    "fontSize": 9,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__Patrick Gräser",
    "label": "Patrick Gräser",
    "left": -24,
    "rectBottom": 36,
    "rectLeft": -48,
    "rectRight": 12,
    "rectTop": 46,
    "right": 6,
    "rotate": false,
    "sentimentScore": 100,
    "square": 120,
    "top": 23,
  },
  Object {
    "bottom": -64,
    "color": "rgb(155, 155, 155)",
    "fontSize": 8,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__Panorama Bar Music",
    "label": "Panorama Bar Music",
    "left": -28,
    "rectBottom": -128,
    "rectLeft": -58,
    "rectRight": -50,
    "rectTop": -50,
    "right": -26,
    "rotate": true,
    "sentimentScore": 50,
    "square": 117,
    "top": -26,
  },
  Object {
    "bottom": 23,
    "color": "rgb(155, 155, 155)",
    "fontSize": 9,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__Germany",
    "label": "Germany",
    "left": -1,
    "rectBottom": 44,
    "rectLeft": -2,
    "rectRight": 38,
    "rectTop": 54,
    "right": 19,
    "rotate": false,
    "sentimentScore": 80,
    "square": 100,
    "top": 27,
  },
  Object {
    "bottom": -31,
    "color": "rgb(155, 155, 155)",
    "fontSize": 9,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__Amsterdam",
    "label": "Amsterdam",
    "left": 23,
    "rectBottom": -62,
    "rectLeft": 42,
    "rectRight": 52,
    "rectTop": -12,
    "right": 26,
    "rotate": true,
    "sentimentScore": 91,
    "square": 100,
    "top": -7,
  },
  Object {
    "bottom": -10,
    "color": "rgb(155, 155, 155)",
    "fontSize": 8,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__Terrace Sundae",
    "label": "Terrace Sundae",
    "left": -18,
    "rectBottom": -22,
    "rectLeft": -36,
    "rectRight": 24,
    "rectTop": -14,
    "right": 12,
    "rotate": false,
    "sentimentScore": 50,
    "square": 90,
    "top": -8,
  },
  Object {
    "bottom": 12,
    "color": "rgb(155, 155, 155)",
    "fontSize": 8,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__Ostgut Ton",
    "label": "Ostgut Ton",
    "left": 20,
    "rectBottom": 22,
    "rectLeft": 38,
    "rectRight": 46,
    "rectTop": 64,
    "right": 23,
    "rotate": true,
    "sentimentScore": 58,
    "square": 84,
    "top": 32,
  },
  Object {
    "bottom": -14,
    "color": "rgb(155, 155, 155)",
    "fontSize": 8,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__Watergate",
    "label": "Watergate",
    "left": -19,
    "rectBottom": -28,
    "rectLeft": -38,
    "rectRight": 2,
    "rectTop": -20,
    "right": 1,
    "rotate": false,
    "sentimentScore": 64,
    "square": 80,
    "top": -11,
  },
  Object {
    "bottom": 11,
    "color": "rgb(155, 155, 155)",
    "fontSize": 9,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__debut LP",
    "label": "debut LP",
    "left": 24,
    "rectBottom": 20,
    "rectLeft": 44,
    "rectRight": 54,
    "rectTop": 58,
    "right": 27,
    "rotate": true,
    "sentimentScore": 100,
    "square": 76,
    "top": 29,
  },
  Object {
    "bottom": -17,
    "color": "rgb(155, 155, 155)",
    "fontSize": 8,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__D/B Presents",
    "label": "D/B Presents",
    "left": -18,
    "rectBottom": -36,
    "rectLeft": -36,
    "rectRight": 14,
    "rectTop": -28,
    "right": 7,
    "rotate": false,
    "sentimentScore": 50,
    "square": 75,
    "top": -15,
  },
  Object {
    "bottom": -54,
    "color": "rgb(155, 155, 155)",
    "fontSize": 8,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__club culture",
    "label": "club culture",
    "left": -15,
    "rectBottom": -108,
    "rectLeft": -32,
    "rectRight": -24,
    "rectTop": -62,
    "right": -13,
    "rotate": true,
    "sentimentScore": 50,
    "square": 69,
    "top": -32,
  },
  Object {
    "bottom": -21,
    "color": "rgb(155, 155, 155)",
    "fontSize": 9,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__Live set",
    "label": "Live set",
    "left": -13,
    "rectBottom": -44,
    "rectLeft": -26,
    "rectRight": 6,
    "rectTop": -34,
    "right": 3,
    "rotate": false,
    "sentimentScore": 75,
    "square": 64,
    "top": -18,
  },
  Object {
    "bottom": -66,
    "color": "rgb(155, 155, 155)",
    "fontSize": 7,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__Hammered",
    "label": "Hammered",
    "left": -13,
    "rectBottom": -132,
    "rectLeft": -28,
    "rectRight": -20,
    "rectTop": -94,
    "right": -11,
    "rotate": true,
    "sentimentScore": 20,
    "square": 57,
    "top": -48,
  },
  Object {
    "bottom": -27,
    "color": "rgb(155, 155, 155)",
    "fontSize": 8,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__Barcelona",
    "label": "Barcelona",
    "left": -19,
    "rectBottom": -56,
    "rectLeft": -38,
    "rectRight": 0,
    "rectTop": -48,
    "right": -1,
    "rotate": false,
    "sentimentScore": 50,
    "square": 57,
    "top": -25,
  },
  Object {
    "bottom": -46,
    "color": "rgb(155, 155, 155)",
    "fontSize": 8,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__Ben Klock",
    "label": "Ben Klock",
    "left": -12,
    "rectBottom": -92,
    "rectLeft": -26,
    "rectRight": -18,
    "rectTop": -54,
    "right": -10,
    "rotate": true,
    "sentimentScore": 50,
    "square": 57,
    "top": -28,
  },
  Object {
    "bottom": -33,
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__dance music",
    "label": "dance music",
    "left": -10,
    "rectBottom": -66,
    "rectLeft": -20,
    "rectRight": 16,
    "rectTop": -60,
    "right": 8,
    "rotate": false,
    "sentimentScore": 0,
    "square": 54,
    "top": -31,
  },
  Object {
    "bottom": -64,
    "color": "rgb(155, 155, 155)",
    "fontSize": 8,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__Odessa",
    "label": "Odessa",
    "left": -11,
    "rectBottom": -128,
    "rectLeft": -24,
    "rectRight": -16,
    "rectTop": -100,
    "right": -9,
    "rotate": true,
    "sentimentScore": 65,
    "square": 42,
    "top": -51,
  },
  Object {
    "bottom": -20,
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__London",
    "label": "London",
    "left": 4,
    "rectBottom": -40,
    "rectLeft": 6,
    "rectRight": 28,
    "rectTop": -34,
    "right": 14,
    "rotate": false,
    "sentimentScore": 1,
    "square": 33,
    "top": -18,
  },
  Object {
    "bottom": -18,
    "color": "rgb(155, 155, 155)",
    "fontSize": 8,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__Disco",
    "label": "Disco",
    "left": -24,
    "rectBottom": -36,
    "rectLeft": -50,
    "rectRight": -42,
    "rectTop": -14,
    "right": -22,
    "rotate": true,
    "sentimentScore": 50,
    "square": 33,
    "top": -8,
  },
  Object {
    "bottom": -31,
    "color": "rgb(155, 155, 155)",
    "fontSize": 8,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__Code",
    "label": "Code",
    "left": -7,
    "rectBottom": -64,
    "rectLeft": -14,
    "rectRight": 6,
    "rectTop": -56,
    "right": 3,
    "rotate": false,
    "sentimentScore": 68,
    "square": 30,
    "top": -29,
  },
  Object {
    "bottom": -24,
    "color": "rgb(155, 155, 155)",
    "fontSize": 8,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__Jun",
    "label": "Jun",
    "left": -18,
    "rectBottom": -48,
    "rectLeft": -36,
    "rectRight": -28,
    "rectTop": -36,
    "right": -15,
    "rotate": true,
    "sentimentScore": 50,
    "square": 24,
    "top": -19,
  },
  Object {
    "bottom": -10,
    "color": "rgb(155, 155, 155)",
    "fontSize": 8,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__UK",
    "label": "UK",
    "left": 13,
    "rectBottom": -22,
    "rectLeft": 24,
    "rectRight": 36,
    "rectTop": -14,
    "right": 18,
    "rotate": false,
    "sentimentScore": 50,
    "square": 18,
    "top": -8,
  },
  Object {
    "bottom": 4,
    "color": "rgb(155, 155, 155)",
    "fontSize": 8,
    "glyphsXOffset": 0,
    "glyphsYOffset": 0,
    "id": "1751295897__DJ",
    "label": "DJ",
    "left": -19,
    "rectBottom": 6,
    "rectLeft": -38,
    "rectRight": -30,
    "rectTop": 14,
    "right": -16,
    "rotate": true,
    "sentimentScore": 54,
    "square": 16,
    "top": 7,
  },
]
`;

exports[`positioningAlgorithm tests prepareData tests should return correct result 1`] = `
Array [
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__Odessa",
    "label": "Odessa",
    "sentimentScore": 65,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__DJ",
    "label": "DJ",
    "sentimentScore": 54,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__Ostgut Ton",
    "label": "Ostgut Ton",
    "sentimentScore": 58,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__Hammered",
    "label": "Hammered",
    "sentimentScore": 20,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__Code",
    "label": "Code",
    "sentimentScore": 68,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__Quantified Drunk",
    "label": "Quantified Drunk",
    "sentimentScore": 50,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__Berghain resident",
    "label": "Berghain resident",
    "sentimentScore": 73,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__San Soda's Panorama Bar",
    "label": "San Soda's Panorama Bar",
    "sentimentScore": 50,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__Germany",
    "label": "Germany",
    "sentimentScore": 80,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__Amsterdam",
    "label": "Amsterdam",
    "sentimentScore": 91,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__Kantine am Berghain",
    "label": "Kantine am Berghain",
    "sentimentScore": 59,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__London",
    "label": "London",
    "sentimentScore": 1,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__UK",
    "label": "UK",
    "sentimentScore": 50,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__Marcel Dettmann",
    "label": "Marcel Dettmann",
    "sentimentScore": 87,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__Disco",
    "label": "Disco",
    "sentimentScore": 50,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__Barcelona",
    "label": "Barcelona",
    "sentimentScore": 50,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__Watergate",
    "label": "Watergate",
    "sentimentScore": 64,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__debut LP",
    "label": "debut LP",
    "sentimentScore": 100,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__Patrick Gräser",
    "label": "Patrick Gräser",
    "sentimentScore": 100,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__Panorama Bar in Berlin",
    "label": "Panorama Bar in Berlin",
    "sentimentScore": 83,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__legendary nightclub",
    "label": "legendary nightclub",
    "sentimentScore": 150,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__Ben Klock",
    "label": "Ben Klock",
    "sentimentScore": 50,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 36,
    "id": "1751295897__Mixes",
    "label": "Mixes",
    "sentimentScore": 895,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__Panorama Bar Music",
    "label": "Panorama Bar Music",
    "sentimentScore": 50,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__Terrace Sundae",
    "label": "Terrace Sundae",
    "sentimentScore": 50,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__Jun",
    "label": "Jun",
    "sentimentScore": 50,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__Live set",
    "label": "Live set",
    "sentimentScore": 75,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__dance music",
    "label": "dance music",
    "sentimentScore": 0,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__club culture",
    "label": "club culture",
    "sentimentScore": 50,
  },
  Object {
    "color": "rgb(155, 155, 155)",
    "fontSize": 6,
    "id": "1751295897__D/B Presents",
    "label": "D/B Presents",
    "sentimentScore": 50,
  },
]
`;
